### DISCUSSION TITLE
Expandable muti-bit inputs/outputs

### DISCUSSION THREAD

**Discussion by richardabendroth** [flandreas/antares#191] [Ideas]
It would be a handy feature to have Circuits with i.e. 4-Bit inputs to feature an option at the inputs to automatically expand those to a combiner (or splitter for outputs respectively) so the user does not need to place a splitter/combiner manually. This makes a lot of sense for ICs where you would like to group the D_IN and D_OUT so the learner has it easier to understand the meaning of the pins while still being practical for connecting those to individual wires when needed.

**Comment by flandreas** [comment]
Can you elaborate a bit more? How would the component look like if you choose this option for a particular input? Would there a combiner hanging right at the pin of this input? Or would the pin show a popup when the user draws a wire to this input, asking him which of the 4 bits he wants to connect? Would this then graphically look like 4 wires be connected to a single pin?

**Comment by richardabendroth** [comment]
I like the 2nd idea a lot. Like you said, the individual wires could just end up in the bigger pin. Or the pin could have a little square on it that denotes that. When clicking or hovering the square (or whatever looks good) would then show an overlay visualizing which wire goes to which of the x pins. This could also be used for editing. So when pulling a wire onto the square (with mouse still down) the overlay opens and the user can connect to the individual pins within the pin bundle.

**Comment by richardabendroth** [comment]
Maybe like this?

![image](https://user-images.githubusercontent.com/25486087/119165595-f16b5480-ba5d-11eb-8454-d2b5d4320ae9.png)
![image](https://user-images.githubusercontent.com/25486087/119165615-fa5c2600-ba5d-11eb-9ff8-fe2651b338a1.png)
![image](https://user-images.githubusercontent.com/25486087/119165645-00ea9d80-ba5e-11eb-919a-b1352e54ce77.png)
![image](https://user-images.githubusercontent.com/25486087/119165797-2a0b2e00-ba5e-11eb-8c15-6ded5e63d934.png)
![image](https://user-images.githubusercontent.com/25486087/119165841-38f1e080-ba5e-11eb-9ffe-eaab63a62e36.png)

**Comment by flandreas** [comment]
Thank you for your detailed design proposal, that's impressive! However, I still feel a little uncomfortable with this feature. First, it would be quite hard to implement, as it breaks a couple of existing invariants, such as a pin has at most 1 connected wire on the view layer. I also doubt that we could find a solution for how to layout multiple wires arriving at the same pin that doesn't confuse the user.

Second, I looked around to see if other tools contain such a feature, and it couldn't find one. In terms of requirement engineering, I would now ask you for a concrete circuit were you really miss this feature. I went over the "Tanenbaum Example" and couldn't find a case where this feature would have been beneficial.

Maybe your requirement arises from the fact that the Antares standard library contains a lot of components with multi-bit pins? Maybe this is wrong, and multi-bit pins should only be used when it is quite sure that the user will bring a wire with the same bit width? And if a component library mainly contains components with lots of single-bit pins, we should have more convenient ways to easily concentrate single-bit-pins right at the pins of a component? And this is a feature that I DID see in other tools..

**Comment by Kitick** [comment]
For me, what I understand is that you want it to basically auto create splitters or mergers so that you don't have to have extra components in the circuit. But in my opinion, I'd think that it would be pretty difficult to implement because of how the current way the wires and pins work. But also to the fact that in my circuits, I've only ever needed splitters or whatnot for I/O. And honestly, set the prob delay to 0, and its like its not there. So I agree with flanderas that while it may be an interesting feature, I don't see many use cases and/or the need for it over just simply adding a extra splitter every so often.

I also agree that a better solution would be to just have more components that have more bit width options. I'd say at least in powers of two up to 8. So it would have 1, 2, 4, and 8 bit width variants. Or, just have 1 and 4. and if you need 2, cascade two 1s, and 8 is two 4s. (However 8 is a nice number for data transmission)
